### PR TITLE
Significant reorganization.

### PR DESCRIPTION
**Pull request by elibixby**
@puneith Please comment on any changes that are confusing.

I'm still (non deterministically) running into non-elegant coordinator exists. We may have to hardcode `tensorflow.python.framework.errors_impl.CancelledError` as a safe exception for the coordinator.

Also, I haven't figured out how to zero out accuracy between evaluations... So it's a running total accuracy for the entire training job. 

NOTE: I may have to go back to averaging accuracies in python rather than using streaming_accuracy.

### CODE REVIEW

File: census/vanilla-tf/trainer/task.py [hunk 1]
```diff
@@ -44,228 +43,149 @@
...
-DEFAULTS = [[0], [''], [0], [''], [0], [''], [''], [''], [''], [''],
-                                    [0], [0], [0], [''], ['']]
+CSV_COLUMN_DEFAULTS = [[0.], [''], [0.], [''], [0.], [''], [''], [''], [''], [''],
+                                    [0.], [0.], [0.], [''], ['']]
 
 # Categorical columns with vocab size
-CATEGORICAL_COLS = (('gender', 2), ('race', 5), ('education', 16), ('marital_status', 7),
+HASH_BUCKET_COLS = (('education', 16), ('marital_status', 7),
                     ('relationship', 6), ('workclass', 9),('occupation', 15), ('native_country', 42))
+KEY_COLS = (('gender', ('female', 'male')), ('race', ('Amer-Indian-Eskimo',
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: What are we achieving by explicitly mentioning "male" and "female" vs saying 2 and then use that fast hashing.

File: census/vanilla-tf/trainer/model.py [hunk 2]
```diff
@@ -46,22 +30,27 @@ def inference(input_x, hidden_units=[100,70,50,25], num_classes=2):
...
+    The last layer values
   """
-  layers_size = [input_x.get_shape()[1]] + hidden_units
-  layers_shape = zip(layers_size[0:],layers_size[1:])
+  print(inputs.get_shape())
+  layers_size = [inputs.get_shape()[1]] + hidden_units
+  layers_shape = zip(layers_size[0:],layers_size[1:] + [num_classes])
+
+  curr_layer = inputs
+  with tf.variable_scope('dnn',
```
Comment (puneith) [inline census/vanilla-tf/trainer/model.py]: So I had it like this before but Nikhil made a comment on separating hidden layer vs output layer. I think that comment makes sense and so we should show the distinction in creating the network (hidden + output).

File: census/vanilla-tf/trainer/model.py [hunk 3]
```diff
@@ -46,22 +30,27 @@ def inference(input_x, hidden_units=[100,70,50,25], num_classes=2):
...
+        biases = tf.get_variable(
+            'biases', shape[1], initializer=tf.zeros_initializer(tf.float32))
+
+      curr_layer = tf.nn.relu(tf.matmul(curr_layer, weights) + biases)
 
-  curr_layer = input_x
+    return curr_layer
 
-  # Creates the relu hidden layers
-  for num, shape in enumerate(layers_shape):
```
Comment (puneith) [inline census/vanilla-tf/trainer/model.py]: Same comment as above. There is merit in keeping this.

File: census/vanilla-tf/trainer/task.py [hunk 4]
```diff
@@ -44,228 +43,149 @@
...
+
+  return train_op, accuracy_op, update_acc, global_step, predictions
+
+
+def convert_sparse_columns(features):
+  for col, bucket_size in HASH_BUCKET_COLS:
+    features[col] = string_ops.string_to_hash_bucket_fast(
+            features[col], bucket_size)
+  
+  for col, keys in KEY_COLS:
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: I think re: comment above this can be merged with the "in HASH_BUCKET_COLS" code block.

File: census/vanilla-tf/trainer/task.py [hunk 5]
```diff
@@ -44,228 +43,149 @@
...
+    table = tf.contrib.lookup.string_to_index_table_from_tensor(
+        tf.constant(keys))
+    features[col] = table.lookup(features[col])
+  
+  for col, size in CATEGORICAL_COLS:
+    features[col] = tf.squeeze(tf.one_hot(
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: Why do you need to squeeze it?

File: census/vanilla-tf/trainer/task.py [hunk 6]
```diff
@@ -274,82 +194,98 @@ def parse_tf_config():
...
+                           task_index=task_index)
 
-  reader = tf.TextLineReader(skip_header_lines=skiprows)
-  key, value = reader.read(filename_queue)
-  value_columns = tf.expand_dims(value, -1)
+  if job_name == 'ps':
+    server.join()
+    return
+  elif job_name in ['master', 'worker']:
+    return run(server.target, job_name == 'master')
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: This should be job_name not job_name == 'master'

File: census/vanilla-tf/trainer/task.py [hunk 7]
```diff
@@ -274,82 +194,98 @@ def parse_tf_config():
...
 
-  columns = tf.decode_csv(
-      value_columns,record_defaults=DEFAULTS)
 
+def parse_label_column(label_string_tensor): 
+  """Parses a string tensor into the label tensor
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: Why do you need the lookup logic  (I mean function parse_label_column) Simply do tf.to_int32

File: census/vanilla-tf/trainer/task.py [hunk 8]
```diff
@@ -274,82 +194,98 @@ def parse_tf_config():
...
+  if shuffle:
+    features = tf.train.shuffle_batch(
+        features,
+        batch_size,
+        capacity=batch_size * 10,
+        min_after_dequeue=batch_size*2 + 1,
+        num_threads=multiprocessing.cpu_count(),
+        enqueue_many=True,
+    )
+  label_tensor = parse_label_column(features.pop(LABEL_COLUMN))
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: See comment above. Don't need the function parse_label_column.

File: census/vanilla-tf/trainer/task.py [hunk 9]
```diff
@@ -44,228 +43,149 @@
...
-      graph = make_graph(inputs, labels)
-
-    init = tf.global_variables_initializer()
-
-    with tf.train.MonitoredTrainingSession(master=server.target,
+      mode = tf.placeholder(shape=[], dtype=tf.string)
+      eval_features, eval_label = input_fn(
+          eval_data_paths, shuffle=False, batch_size=batch_size)
+      train_features, train_label = input_fn(
+          train_data_paths, shuffle=True, num_epochs=num_epochs, batch_size=batch_size)
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: In this logic between lines 137-154 can we not create two graphs one for training and one for eval directly vs doing this cond thing?

File: census/vanilla-tf/trainer/task.py [hunk 10]
```diff
@@ -44,228 +43,149 @@
...
+              lambda: [train_features[k] for k in sorted_keys],
+              lambda: [eval_features[k] for k in sorted_keys]
+          )
+      ))
+      labels = tf.cond(is_train, lambda: train_label, lambda: eval_label)
+      train_op, accuracy_op, eval_op, global_step_tensor, predictions = model_fn(
+          inputs, labels, learning_rate=learning_rate)
+
+    with tf.train.MonitoredTrainingSession(master=target,
                                            is_chief=is_chief,
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: Where is is_chief computed? I guess I am not finding it.

File: census/vanilla-tf/trainer/task.py [hunk 11]
```diff
@@ -44,228 +43,149 @@
...
+        while step < max_steps and not coord.should_stop():
+            if is_chief and step - last_eval > eval_every:
+                last_eval = step
+                for _ in range(eval_steps):
+                    session.run([eval_op], feed_dict={mode: EVAL})
+                accuracy = session.run([accuracy_op], feed_dict={mode: EVAL})
+                print("Accuracy at step: {} is {}".format(step, accuracy))
+            step, _ = session.run(
+                [global_step_tensor, train_op],
+                feed_dict={mode: TRAIN}
```
Comment (puneith) [inline census/vanilla-tf/trainer/task.py]: Re: comment above can we get rid of mode and create training and eval graphs. Possible or am I missing some logic which is required?

File: census/vanilla-tf/trainer/task.py (hunk 1 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: We're demonstrating using explicit keys for hash buckets. We can remove since we demonstrate this with the label column. But this gives us some sort of parity with the tf.learn example.

File: census/vanilla-tf/trainer/model.py (hunk 2 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/model.py]: If we want to demonstrate output, we can move the softmax from the logit layers here and return `logits, predictions` as a tuple. However note that predictions is currently unused, so this will go unevaluated except in the predict case. `sparse_softmax_cross_entropy_with_logits` shouldn't take inputs from a softmax layer (or doesn't need to rather).

File: census/vanilla-tf/trainer/task.py (hunk 5 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: Concat requires Tensors all of the same rank, and we don't transform the real valued columns.

File: census/vanilla-tf/trainer/task.py (hunk 4 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: This applies to all the categorical columns. Although yes we could clean up if we eliminate the key columns.

File: census/vanilla-tf/trainer/task.py (hunk 7 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: This supports moving to >2 labels much easier (which the team wanted since there will be a guide for "Adapting this model to your own data).

File: census/vanilla-tf/trainer/task.py (hunk 6 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: No it shouldn't `run` takes `is_chief` not `job_name`. I'll change these to kwargs to make that clearer.

File: census/vanilla-tf/trainer/task.py (hunk 11 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: So, this was the other option I considered, but I went this direction because to move weights between the two graphs, you'll need to load a checkpoint, which ties evaluation frequency to checkpointing frequency. Which is a frustration I've heard from several tf.learn users. 

However, the benefit of evaluating with a fresh session/graph every time is that you will use the same data over and over again. (Although you'll incur I/O costs every time which seems silly).

That said, if `eval_steps` and `batch_size` is high enough you the Law of Large Numbers should kick in and that shouldn't be a problem, so I think this is the better solution. Especially since it prevents potentially "getting lucky" with your evaluation dataset.

File: census/vanilla-tf/trainer/task.py (hunk 1 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: Ok

File: census/vanilla-tf/trainer/task.py (hunk 6 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: Sorry read the code wrong. Too tired I guess. Yes correct.

File: census/vanilla-tf/trainer/task.py (hunk 10 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: nvm about this. addresed above.

File: census/vanilla-tf/trainer/task.py (hunk 11 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: But we can share weights even on dist setting right? So why do we need to load checkpoint?

File: census/vanilla-tf/trainer/task.py (hunk 7 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: Sorry - But when you do one hot life is good. Why lookup stuff and add more lines?

File: census/vanilla-tf/trainer/model.py (hunk 2 above)
Comment (puneith) [reply census/vanilla-tf/trainer/model.py]: So here we are adding the relu in output layer when it should be softmax.

File: census/vanilla-tf/trainer/task.py (hunk 4 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: I think we can unify them and reduce lines of code.

File: census/vanilla-tf/trainer/task.py (hunk 7 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: Uhhh you need to do the lookup to convert string labels to integer labels. Then you STILL need to do one_hot.

File: census/vanilla-tf/trainer/task.py (hunk 11 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: I see what you're suggesting. If we had a dedicated evaluation master that would work. Just only run the evaluation OP on the evaluation master. But we need this to work in single worker mode, meaning we need a single task to be able to do both training and evaluation. Single task = single graph OR switching between graphs. Switching between graphs you'd need to load the parameters from checkpoints (this is how tf.learn does it).

@brandondutra for comment on this approach.

File: census/vanilla-tf/trainer/model.py (hunk 2 above)
Comment (elibixby) [reply census/vanilla-tf/trainer/model.py]: Again no. Why should it be softmax? I think I explained this to you in chat?

File: census/vanilla-tf/trainer/task.py (hunk 11 above)
Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: But the more common use case where we expect performance to be a concern will actually be address by distributed. Why would anyone use single node and expect performance to be pretty good?

Comment (elibixby) [reply census/vanilla-tf/trainer/task.py]: @puneith No it wouldn't be addressed by distributed. What you're proposing only works if you have a dedicated evaluation master. Otherwise you need a way to have continuous evaluation in one task. This requires either loading from checkpoints (tying to check-pointing) or merging the two graphs.

Comment (puneith) [reply census/vanilla-tf/trainer/task.py]: Sure and why isn't that distributed and preferred? Are you concerned that having two master will make the design complicated?
